src/iteration.jl

```diff
@@ -162,48 +160,34 @@ end
 """
     Base.getindex(ba::LazyBranch{T, J}, idx::Integer) where {T, J}
 
-Get the `idx`-th element of a `LazyBranch`, starting at `1`. If `idx` is
-within the range of `ba.buffer_range`, it will directly return from `ba.buffer`.
-If not within buffer, it will fetch the correct basket by calling [`basketarray`](@ref)
-and update buffer and buffer range accordingly.
-
-!!! warning
-    Because currently we only cache a single basket inside `LazyBranch` at any given
-    moment, access a `LazyBranch` from different threads at the same time can cause
-    performance issue and incorrect event result.
+Get the `idx`-th element of a `LazyBranch`, starting at `1`.
 """
 
 function Base.getindex(ba::LazyBranch{T,J,B}, idx::Integer) where {T,J,B}
-    tid = Threads.threadid()
-    br = @inbounds ba.buffer_range[tid]
-    # index within the basket
-    localidx = if idx ∉ br
-        _localindex_newbasket!(ba, idx, tid)
-    else
-        idx - br.start + 1
-    end
-    return @inbounds ba.buffer[tid][localidx]
+    # cached get the cluster this idx belongs to
+    br, basket_data = _getbasket(ba, idx) 
+    localidx = idx - first(br) + 1
+    return basket_data[localidx]
 end
 
-function _localindex_newbasket!(ba::LazyBranch{T,J,B}, idx::Integer, tid::Int) where {T,J,B}
+function _getbasket(ba::LazyBranch{T,J,B}, idx::Integer) where {T,J,B}
     seek_idx = findfirst(x -> x > (idx - 1), ba.fEntry) #support 1.0 syntax
-    br = _get_buffer_range(ba, tid, seek_idx)
-    ba.buffer_range[tid] = br
-    return idx - br.start + 1
-end
-
-function _get_buffer_range(ba::LazyBranch{T, J, B}, tid::Integer, seek_idx::Integer) where {T,J,B}
     seek_idx -= 1
-    ba.buffer[tid] = basketarray(ba.f, ba.b, seek_idx)
-    (ba.fEntry[seek_idx] + 1)::Int:(ba.fEntry[seek_idx + 1])::Int
-end
+    lru = ba.buffers
+    br = (ba.fEntry[seek_idx] + 1)::Int:(ba.fEntry[seek_idx + 1])::Int
+    data = get!(lru, br) do
+        basketarray(ba.f, ba.b, seek_idx)
+    end
 
-function _get_buffer_range(ba::LazyBranch{T, J, B}, tid::Integer, ::Nothing) where {T,J,B}
-    ba.buffer[tid] = basketarray(ba.f, ba.b, -1)  # -1 indicating recovered basket mechanics
-    # FIXME: this range is probably wrong for jagged data with non-empty offsets
-    (ba.b.fBasketEntry[end] + 1)::Int:ba.b.fEntries::Int
+    return br, data
 end
 
+# function _get_buffer_range(ba::LazyBranch{T, J, B}, tid::Integer, ::Nothing) where {T,J,B}
```

Comment:
@tamasgal do you remember what is this for? didn't hit in test it seems

Reply:
We refactored a few things, not sure about the origins 🙈

Reply:
https://github.com/JuliaHEP/UnROOT.jl/commit/81007e16fe0253e4d9e6652f1cf41ae853c9061b

it's for recovered basket when `seek_idx` is nothing